onnxscript/rewriter/fuse_relus_clips_test.py

```diff
@@ -0,0 +1,320 @@
+                """
+                    x1 = Relu(X)
+                    Y = Clip(x1, min)
+                """,
+                fuse_successive_clip_relu_rule,
+            ),
+            (
+                "clip_then_relu",
+                """
+                    x1 = Clip(X, min)
+                    Y = Relu(x1)
+                """,
+                fuse_successive_relu_clip_rule,
+            ),
+        ]
+    )
+    def test_fuse_successive_relu_clip_non_initializers(self, _, nodes, rewrite_rule):
+        model = ir.from_onnx_text(f"""
+            < ir_version: 10, opset_import: ["" : 20] >
+            test_model (float[N, 32, 14] X) => (float [N, ?, ?] Y)
+            {{
+                min = ReduceMean<keepdims=0>(X)
+                {nodes}
+            }}
+        """)
+        self.run_failed_condition_test(model, rewrite_rule, "is not a constant initializer.")
+
+    @parameterized.parameterized.expand(
+        [
+            (
+                "relu_then_clip",
+                """
+                    x1 = Relu(X)
+                    Y = Clip(x1, min)
+                """,
+                fuse_successive_clip_relu_rule,
+            ),
+            (
+                "clip_then_relu",
+                """
+                    x1 = Clip(X, min)
+                    Y = Relu(x1)
+                """,
+                fuse_successive_relu_clip_rule,
+            ),
+        ]
+    )
+    def test_fuse_successive_relu_clip_graph_inputs(self, _, nodes, rewrite_rule):
+        model = ir.from_onnx_text(f"""
+            < ir_version: 10, opset_import: ["" : 20] >
+            test_model (float[N, 32, 14] X, float min) => (float [N, ?, ?] Y)
+            {{
+                {nodes}
+            }}
+        """)
+        self.run_failed_condition_test(model, rewrite_rule, "is a graph input.")
+
+
+class FuseSuccessiveClipTest(_FuseReluClipTestBase):
+    @parameterized.parameterized.expand(
+        [
+            ("float", "float"),
+            ("int32", "int32"),
+        ]
+    )
+    def test_fuse_succesive_clips(self, _, dtype):
+        model = ir.from_onnx_text(f"""
+            < ir_version: 10, opset_import: ["" : 20] >
+            test_model ({dtype}[N, 32, 14] X) => ({dtype} [N, ?, ?] Y)
+            <{dtype} min1 = {{0}}, {dtype} max1 = {{4}},
+            {dtype} min2 = {{1}}, {dtype} max2 = {{11}},
+            {dtype} min3 = {{3}}, {dtype} max3 = {{7}}>
+            {{
+                x1 = Clip(X, min1, max1)
+                x2 = Clip(x1, min2, max2)
+                Y  = Clip(x2, min3, max3)
+            }}
+        """)
+        self.run_test(model, expected_op_type="Clip", dtype=dtype)
+
+    def test_fuse_succesive_clips_no_min(self):
+        model = ir.from_onnx_text("""
+            < ir_version: 10, opset_import: ["" : 20] >
+            test_model (float[N, 32, 14] X) => (float [N, ?, ?] Y)
+            <float max1 = {4.0}, float max2 = {7.0}>
+            {
+                x1 = Clip(X,, max1)
```

Comment:
@justinchuby I didn't know ONNX op accepts None as input. What is this in proto and IR (the min input)?

Reply:
Yeah, I see what we did in aten_clamp.

Reply:
It's surprising that we can provide the second one and leave the first one blank though.